File: C++ Graphs/graph.cpp
#ifndef GRAPH_H
#define GRAPH_H

#include <vector>
#include <iostream>

class node {
    public:
        //Assume the data in our graph 
        //is just the nodes location on 
        //a 2d plane (assume a space of
        //1000, 1000 pixels)
        int location[2];

        node(int x, int y) {
            location[0] = x;
            location[1] = y;
        }

    friend class graph;
};

class graph {
    public:

    std::vector<node> allNodes;
    std::vector<std::vector<bool>> adjMat;

    //Constructor
    graph(node start) {
        allNodes.push_back(start);

        std::vector<bool> temp = {false};
        adjMat.push_back(temp);
    }

    //Returns whether or not a node is in the graph
    bool inGraph(node input) {

        for(int i = 0; i < allNodes.size(); i++) {
            
            if(input.location[0] == allNodes[i].location[0] && input.location[1] == allNodes[i].location[1]){
                return true;
            }
        }

        return false;
    }

    void addNode(int inputX, int inputY) {

        //Create node with given values
        node input(inputX, inputY);
        //Add node to graph
        allNodes.push_back(input);

        //Expand each ROW of the matrix to include new node
        for(int i = 0; i < adjMat.size(); i++) {
            adjMat[i].push_back(false);
        }

        //Add an additional row to the 
        //matrix to include the new node
        std::vector<bool> temp;
        for(int i = 0; i <= adjMat.size(); i++) {
            temp.push_back(false);
        }

        adjMat.push_back(temp);
    }

    void deleteNode(node input);
    
    void addConnection(node one, node two) {

        //Determines if these nodes are even in the graph
        if(inGraph(one) && inGraph(two)) {

            //Find where they are in our allNodes vector
            int onePosition, twoPosition;
            for(int i = 0; i < allNodes.size(); i++) {
                if(one.location[0] == allNodes[i].location[0] && one.location[1] == allNodes[i].location[1]) {
                    onePosition = i;
                }

                if(two.location[0] == allNodes[i].location[0] && two.location[1] == allNodes[i].location[1]) {
                    twoPosition = i;
                }
            }

            adjMat[onePosition][twoPosition] = 1;
            adjMat[twoPosition][onePosition] = 1;
        }
        
    }

    void deleteConnection(node one, node two);

    void showNodes(){

        for(int i = 0; i < allNodes.size(); i++) {
            std::cout << "(" << allNodes[i].location[0] << ", " 
                             << allNodes[i].location[1] << ")\n";
        }
    }

    void showMatrix() {

        for(int i = 0; i < adjMat.size(); i++) {
            for(int j = 0; j < adjMat[i].size(); j++) {
                std::cout << adjMat[i][j] << "\t";
            }
            std::cout << "\n";
        }
    }
};

#endif

File: Python Graph/graph.py
class node():
    
    def __init__(self, x, y):
        self.location = [x, y]

class graph():

    #Constructor
    def __init__(self, start):
        self.all_nodes = [start]
        self.adj_mat = [[False]]

    #Returns whether a node is in the graph or not
    def in_graph(self, input):
        
        #Loops through each node in all_nodes
        for node in self.all_nodes:

            #Checks to see if data is the same
            if input.location[0] == node.location[0] and input.location[1] == node.location[1]:
                return True

        return False
    
    #Adds a new node to the graph
    def add_node(self, input_x, input_y):

        #Create a node with input values
        input = node(input_x, input_y)
        #Add node to graph
        self.all_nodes.append(input)

        #Expand each row of the matrix to include the new node
        for i in range(0, len(self.adj_mat)):
            self.adj_mat[i].append(False)

        #Add an additional row to the matrix to include the new node
        temp = []
        for i in range(0, len(self.adj_mat) + 1):
            temp.append(False)
        
        #Add row to matrix
        self.adj_mat.append(temp)
    
    #Adds a connection between two nodes in the graph
    def add_connection(self, first_node, second_node):

        #Determine if the nodes are even in the graph
        if self.in_graph(first_node) and self.in_graph(second_node):

            #Find where nodes are in our all_nodes list
            for i in range(len(self.all_nodes)):

                #If their coord values are the same then we have found it
                if(first_node.location[0] == self.all_nodes[i].location[0] 
                   and first_node.location[1] == self.all_nodes[i].location[1]):
                    pos_one = i

                if(second_node.location[0] == self.all_nodes[i].location[0] 
                   and second_node.location[1] == self.all_nodes[i].location[1]):
                    pos_two = i
            
            self.adj_mat[pos_one][pos_two] = True
            self.adj_mat[pos_two][pos_one] = True

    #Prints out the VALUES of each of the nodes in the graph
    def show_nodes(self):

        for node in self.all_nodes:
            print("(%s,%s)" % (node.location[0], node.location[1]))
    
    #Prints out the adjency matrix
    def show_matrix(self):
        
        for i in range(len(self.adj_mat)):
            for j in range(len(self.adj_mat)):

                print(self.adj_mat[i][j], end=" ")
            print("\n")